docs/customruns.md

```diff
@@ -0,0 +1,413 @@
+   `ApiVersion` and `Kind`. Custom task controllers should handle
+   an empty spec, either with a default behaviour, in a case no default
+   behaviour is supported then, appropriate validation error should be
+   updated to the `CustomRun`'s status.
+
+### Cancellation
+
+Implementing cancellation is now `required` instead of `optional`. If the Custom Task implementor does not support cancellation via `.spec.status`, `Pipeline` timeouts and `PipelineRun` cancellation will not work properly.
+
+Pipeline Controller sets the `spec.Status` and `spec.StatusMessage` to signal `CustomRuns` about the `Cancellation`, while `CustomRun` controller **MUST** update its `status.conditions` as following once noticed the change on `spec.Status`.
+
+| Pipeline Signal| `spec.Status` | `spec.StatusMessage` | `status.Conditions` | 
+|:--- |:--- |:--- |:---|
+|Parent `PipelineRun` is cancelled | `RunCancelled` | `CustomRun cancelled as the PipelineRun it belongs to has been cancelled. `| `Type: Succeeded`<br>`Status:False`<br>`Reason:CustomRunCancelled` |
+|`Pipeline` timeouts | `RunCancelled` | `CustomRun cancelled as the PipelineRun it belongs to has timed out.`| `Type: Succeeded`<br>`Status:False`<br>`Reason:CustomRunCancelled`|
+
+Note: Custom Task implementors **MUST NOT** update the value of `spec.StatusMessage` and `spec.Status`, because Pipeline Controller depends the two fields.
+
+### Specifying `Timeout`
+
+A custom task specification can be created with `Timeout` as follows:
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  generateName: simpleexample
+spec:
+  timeout: 10s # set timeouts here.
+  params:
+    - name: searching
+      value: the purpose of my existence
+  customRef:
+    apiVersion: custom.tekton.dev/v1alpha1
+    kind: Example
+    name: exampleName
+```
+
+Supporting timeouts is optional but recommended.
+
+#### Developer guide for custom controllers supporting `Timeout`
+
+1. Tekton controllers will never directly update the status of the `CustomRun`, 
+   it is the responsibility of the custom task controller to support timeout.
+   If timeout is not supported, it's the responsibility of the custom task
+   controller to reject `CustomRun`s that specify a timeout value.
+2. On a `pipelineRun` or `pipelineTask` timeout, the status of the 
+   `CustomRun.Spec.Status` is updated to `RunCancelled`. The custom task controller
+   MUST cancel the `CustomRun`. Otherwise, pipeline-level `Timeout` and
+   `Cancellation` won't work for the Custom Task.
+3. A Custom Task controller can watch for this status update
+   (i.e. `CustomRun.Spec.Status == RunCancelled`) and or `CustomRun.HasTimedOut()`
+   and take any corresponding actions (i.e. a clean up e.g., cancel a cloud build,
+   stop the waiting timer, tear down the approval listener).
+4. A Custom Task author can differentiate between a `CustomRun` cancelled by an user
+   from a Run cancellation initiated by a PipelineRun by looking at the 
```

Comment:
I'm not sure the custom task controller needs to know why the custom task was canceled (beyond info present in the message) or what canceled it. Is this StatusMessage part of the api?

Reply:
Sorry, I should have been more clear, the field `StatusMessage` is indeed part of the API, but do we make guarantees about what that message will be? I think it's intended to be human readable and there's a reasonable chance we might want to change it. I'd just put here that "StatusMessage displays a human-readable reason for the cancelation."

Reply:
Makes sense! 👍 Even though Pipeline Controller relies on the `spec.StatusMessage` to distinguish the different cancellation, it's an unnecessary info for CustomRun controllers, we shouldn't guarantee what we'd put in this field and Custom Run controller shouldn't rely on this field.